### PR TITLE
filter, action: Add support for tunnel_key

### PR DESCRIPTION
**Pull request by zhiyuan0x**
Add support for tunnel_key

### CODE REVIEW

**Comment by BSWANG** [comment]
LGTM

**Comment by zhiyuan0x** [comment]
@vishvananda @aboch  can this PR be merged? thanks.

File: filter_linux.go [hunk 1]
```diff
@@ -429,6 +430,29 @@ func EncodeActions(attr *nl.RtAttr, actions []Action) error {
...
+				Action: int32(action.Action),
+			}
+			toTcGen(action.Attrs(), &tun.TcGen)
+			aopts.AddRtAttr(nl.TCA_TUNNEL_KEY_PARMS, tun.Serialize())
+			if action.Action == TCA_TUNNEL_KEY_SET {
+				aopts.AddRtAttr(nl.TCA_TUNNEL_KEY_ENC_KEY_ID, htonl(action.KeyID))
+				if v4 := action.SrcAddr.To4(); v4 != nil {
+					aopts.AddRtAttr(nl.TCA_TUNNEL_KEY_ENC_IPV4_SRC, v4[:])
+				} else {
+					aopts.AddRtAttr(nl.TCA_TUNNEL_KEY_ENC_IPV6_SRC, action.SrcAddr.To16()[:])
```
Comment (aboch) [inline filter_linux.go]: Why `To16()` if you already know it is a IPv6 address?
https://golang.org/src/net/ip.go?s=5477:5499#L199

Comment (zhiyuan0x) [reply filter_linux.go]: I have fixed it. thanks.

**Comment by zhiyuan0x** [comment]
@aboch fixed it. thanks.

File: filter_linux.go [hunk 2]
```diff
@@ -429,6 +430,33 @@ func EncodeActions(attr *nl.RtAttr, actions []Action) error {
...
+			tun := nl.TcTunnelKey{
+				Action: int32(action.Action),
+			}
+			toTcGen(action.Attrs(), &tun.TcGen)
+			aopts.AddRtAttr(nl.TCA_TUNNEL_KEY_PARMS, tun.Serialize())
+			if action.Action == TCA_TUNNEL_KEY_SET {
+				aopts.AddRtAttr(nl.TCA_TUNNEL_KEY_ENC_KEY_ID, htonl(action.KeyID))
+				if v4 := action.SrcAddr.To4(); v4 != nil {
+					aopts.AddRtAttr(nl.TCA_TUNNEL_KEY_ENC_IPV4_SRC, v4[:])
+				} else if v6 := action.SrcAddr.To16(); v6 != nil {
```
Comment (aboch) [inline filter_linux.go]: What about here?

File: filter_linux.go [hunk 3]
```diff
@@ -429,6 +430,33 @@ func EncodeActions(attr *nl.RtAttr, actions []Action) error {
...
+					aopts.AddRtAttr(nl.TCA_TUNNEL_KEY_ENC_IPV6_SRC, v6[:])
+				} else {
+					return fmt.Errorf("invalid src addr %s for tunnel_key action", action.SrcAddr)
+				}
+				if v4 := action.DstAddr.To4(); v4 != nil {
+					aopts.AddRtAttr(nl.TCA_TUNNEL_KEY_ENC_IPV4_DST, v4[:])
+				} else if v6 := action.DstAddr.To16(); v6 != nil {
```
Comment (aboch) [inline filter_linux.go]: And here

File: filter_linux.go (hunk 2 above)
Comment (zhiyuan0x) [reply filter_linux.go]: @aboch In golang, ip address is a simple slice and does not have an address family flag. So if IP.To4() return a nil and  IP.To16() reutrn a not nil, It should be ipv6 address.
This is a similar code: https://golang.org/src/net/ip.go#L420 .

Comment (aboch) [reply filter_linux.go]: ok

**Comment by aboch** [comment]
LGTM

**Pull request merged**